lib/dRep/dRepImpl.py

```diff
@@ -80,102 +77,321 @@ def run_dRep(self, ctx, params):
         """
         # ctx is the context object
         # return variables are: output
-        #BEGIN run_dRep
+ 
+        #BEGIN dereplicate
 
-        
-        dprint('ctx:', ctx)
-        dprint('params:', params)
+        dprint('params', run=locals())
 
 
+        # set up globals ds `_globals` for this API-method run
 
+        reset(_globals) # clear all fields but `debug`
+        _globals.update({
+            **self._globals,
+            'workspace_name': params['workspace_name'],
+            'run_dir': os.path.join(self._globals['shared_folder'], str(uuid.uuid4())),
+            'warnings': [],
+            })
+
+        os.mkdir(_globals.run_dir)
+
+
+
+
+        #
+        ##
+        ###
+        ####
+        #####
+
+
+        def simple_return(msg, report_params_kwargs={}):
+            '''
+            Return with a simple message
+            '''
+            report_params = {
+                    'message': msg,
+                    'workspace_name': _globals.workspace_name,
+                    **report_params_kwargs
+                    }
+            report_info = _globals.kbr.create_extended_report(report_params)
+            output = {
+                'report_name': report_info['name'],
+                'report_ref': report_info['ref'],
+            }
+
+            return [output]
+
+
+
+        # 
+        ##
+        ### check unique UPAs
+        #### 
+        #####
+
+        if len(set(params['genomes_refs'])) < len(params['genomes_refs']):
+            msg = 'Removing duplicate input BinnedContigs'
+            logging.warning(msg)
+            _globals.warnings.append(msg) 
+            params['genomes_refs'] = list(dict.fromkeys(params['genomes_refs']))
```

Comment:
Why create an intermediate dictionary here? I would just make this a set since it is semantically a set. `params['genomes_refs'] = set(params['genomes_refs'])`